### PR TITLE
Fix for optimization sequence in scanning toolchain

### PR DESCRIPTION
**Pull request by TobiasSpohn**
## Description
Fixes that the optimizer could not perform optimization sequences other than the default [2, 1], e.g. ["xy","z"]. By changing the checker function of the ConfigOption `ScannerGui.optimizer_plot_dimensions` to a constructor that correctly checks the constraints. Furthermore, the selection of available sequences is now handled by `scanning_optimize_logic`, the GUI only access it.
A signal is added to lay the ground for future implementation of adjusting the optimizer sequence during runtime.

## Motivation and Context
The `ScannerGui.optimizer_plot_dimensions` ConfigOption is there to adjust the optimization sequence to an arbitrary combination of 2D and 1D optimizations within the constraints of the hardware (i.e. number of axes). It should be possible with e.g. 3 Axes to perform the default 2D and 1D optimization, but also any subset of these. E.g. only a 1D or multiple 1D optimizations. Due to a bug in the checker method of the ConfigOption it was not possible to select only 1D optimizations. Furthermore, it was not possible to change `ScanningOptimizeLogic._scan_sequence` once it was stored as a StatusVar and the default fallback for the StatusVar was hardcoded to a sequence of [2,1].

## How Has This Been Tested?
Default config file

## Types of changes
- [x] Bug fix
- [ ] New feature
- [x] Breaking change (Causes existing functionality to not work as expected) -> The ConfigOption moved to the logic

## Checklist:
- [x] My code follows the code style of this project.
- [ ] I have documented my changes in `/docs/changelog.md`.
- [ ] My change requires additional/updated documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added/updated the config example for any module docstrings as necessary.
- [x] I have checked that the change does not contain obvious errors
(syntax, indentation, mutable default values, etc.).
- [x] I have tested my changes using 'Load all modules' on the default dummy configuration. -> POI manager can't load due to #154 
- [ ] All changed Jupyter notebooks have been stripped of their output cells.

### CODE REVIEW

**Comment by TobiasSpohn** [comment]
With Hotfix #156 load all modules completes without errors.

File: docs/changelog.md [hunk 1]
```diff
@@ -7,6 +7,7 @@
   implementing this interface.
 - Add back scan data and back scan settings getter and setter methods
   to `qudi.interface.scanning_probe_interface`.
+- Moved ConfigOption ScannerGui.optimizer_plot_dimensions to qudi.logic.scanning_optimize_logic and optimizer sequence creation is solely handled by logic
```
Comment (timoML) [inline docs/changelog.md:10]: need to mention the new name of the ConfigOption "optimizer_sequence_dimensions"

**Comment by timoML** [comment]
That's definitely an improvement.
While testing I encountered that, currently, we get a lot of StatusVar related errors when configuring a 2-axis scanner_dummy.
Maybe we should put in some work into fixing these.
After StatusVar clearing, the behavior with 2-axis scanners is clearer now.

**Comment by TobiasSpohn** [comment]
I rewrote the image generation of the dummy. The dummy can now have an arbitrary number of axes.

File: src/qudi/hardware/dummy/scanning_probe_dummy.py [hunk 2]
```diff
@@ -583,24 +761,46 @@ def __stop_timer(self):
...
+
+        return scan_vectors
+
+    @_spot_density.constructor
+    def spot_density_constructor(self, spot_density: float) -> float:
+        volume_edges = [abs(pos_range[1] - pos_range[0]) for pos_range in self._position_ranges.values()]
+        volume = 1
+        for edge in volume_edges:
+            volume *= edge
+        if volume * spot_density ** len(self._position_ranges.keys()) > 80e3:
```
Comment (timoML) [inline src/qudi/hardware/dummy/scanning_probe_dummy.py:793]: Maximum spot number should be a hardware constraint.

Comment (TobiasSpohn) [reply src/qudi/hardware/dummy/scanning_probe_dummy.py]: I implemented it as a constraint and therefore subclassed the `ScanConstraint` class. This however results in the maximum spot number check to occur only during `on_activate` of the module and not during the constructor of the `ConfigOption`.

File: src/qudi/logic/scanning_optimize_logic.py [hunk 3]
```diff
@@ -389,3 +446,19 @@ def _set_default_scan_settings(self):
         self._scan_frequency = {ax.name: max(ax.frequency.minimum, ax.frequency.maximum / 100) for ax in axes.values()}
         self._back_scan_resolution = {}
         self._back_scan_frequency = {}
+
+    def _set_scan_sequence(self):
```
Comment (timoML) [inline src/qudi/logic/scanning_optimize_logic.py:450]: Ambigous name of function; it's not a setter. Maybe _set_default_scan_sequence()?

File: src/qudi/gui/scanning/scannergui.py [hunk 4]
```diff
@@ -439,6 +418,22 @@ def _init_static_dockwidgets(self):
         self.tilt_correction_dockwidget.visibilityChanged.connect(self._mw.action_view_tilt_correction.setChecked)
         self._mw.action_view_tilt_correction.triggered[bool].connect(self.tilt_correction_dockwidget.setVisible)
 
+    def _set_optimizer_dockwidget(self):
```
Comment (timoML) [inline src/qudi/gui/scanning/scannergui.py:421]: Ambigous name of function; it's not a setter. Maybe _init_optimizer_dockwidget()?

**Comment by timoML** [comment]
Except for minor notes above, it looks really good now. We still should:
- run an overnight test with continuous optimization to test the changed BlockingQueuedConnection
- maybe run ruff over the files (@prithviulm).

**Comment by prithviulm** [comment]
I made a branch here https://github.com/Ulm-IQO/qudi-core/tree/embracing_ruff for developing the ruff config we intend to use. @TobiasSpohn could you try it out? The config is within pyproject.toml. Installation instructions are in the commit message, see here for quick reference https://github.com/Ulm-IQO/qudi-core/commit/7a38bc10518604a7fc6b29ffe5c2a9d37ff421b9

**Comment by TobiasSpohn** [comment]
As discussed in the software meeting, all debug and `Qt.BlockingQueuedConnection` changes have been reverted and the branch is ready for testing with the fix for scanning. The thread deadlock issue will need some further investigation and some other PR to fix.

**Review by timoML — DISMISSED** [review]
Looks good now, thank you.

File: src/qudi/gui/scanning/scannergui.py (hunk 4 above)
Comment (TobiasSpohn) [reply src/qudi/gui/scanning/scannergui.py]: changed the name

File: src/qudi/logic/scanning_optimize_logic.py (hunk 3 above)
Comment (TobiasSpohn) [reply src/qudi/logic/scanning_optimize_logic.py:450]: changed the function name

**Review by timoML — APPROVED** [review]
Thank you.

**Pull request merged**